extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java

```diff
@@ -0,0 +1,95 @@
+package io.quarkus.agroal.metrics;
+
+import org.eclipse.microprofile.metrics.Counter;
+
+import io.agroal.api.AgroalDataSource;
+import io.agroal.api.AgroalDataSourceMetrics;
+import io.quarkus.arc.Arc;
+
+public class AgroalCounter implements Counter {
+
+    private String dataSourceName;
+    private AgroalDataSource dataSource;
+    private String metric;
+
+    public AgroalCounter() {
+
+    }
+
+    /**
+     * @param dataSourceName Which datasource should be queried for metric
+     * @param metricName Name of the method from DataSource.getMetrics() that should be called to retrieve the particular value.
+     *        This has nothing to do with the metric name from MP Metrics point of view!
+     */
+    public AgroalCounter(String dataSourceName, String metricName) {
+        this.dataSourceName = dataSourceName;
+        this.metric = metricName;
+    }
+
+    public String getDataSourceName() {
+        return dataSourceName;
+    }
+
+    public void setDataSourceName(String dataSourceName) {
+        this.dataSourceName = dataSourceName;
+    }
+
+    private AgroalDataSource getDataSource() {
+        if (dataSource == null) {
```

Comment:
This double-check locking idiom makes me think `dataSource` should be volatile and also be read into a local variable.

Reply:
+1 `dataSource` must be volatile here. 
But I will avoid using synchronization here and go for getting the datasource from an `@PostConstruct` block

Reply:
I guess you're right. I'm not sure if there's any better way to do this, I would need a callback that gets invoked after the Arc container is started, but before any metric requests arrive. I don't want to turn this into a CDI bean just for the purpose of being able to do `@PostConstruct`, due to the performance penalty, and also potentially breaking user applications by unexpectedly introducing some CDI beans that implement `Gauge` or `Counter`. I will fix the synchronization for now, perhaps we will find a better solution where this is not needed.